[JIT] metacompile isinstance checks

eellison opened the pull request:
This is a series of PRs that will allow us to support adding [padding to conv](https://github.com/pytorch/pytorch/pull/22484) and also reduce the friction of adding method overloads that was brought up in  https://github.com/pytorch/pytorch/pull/23266.

This PR only compiles one if branch if the condition is an isinstance check. This is consistent with what mypy does; it does not report errors if a branch can be determined statically to be unreachable.

```
def foo(x):
    # type: (int) -> int
    if isinstance(x, str):
        return x["1"]
    return x + 1

reveal_type(foo) # no error, shows int -> int
``` 


Stack from [ghstack](https://github.com/ezyang/ghstack):
* #23886 [JIT] add support for overloading functions
* **#23885 [JIT] metacompile isinstance checks**

Differential Revision: [D16697092](https://our.internmc.facebook.com/intern/diff/D16697092)

Commit:
metacompile isinstance checks

Commit:
Update on "[JIT] metacompile isinstance checks"

This is a series of PRs that will allow us to support adding [padding to conv](https://github.com/pytorch/pytorch/pull/22484) and also reduce the friction of adding method overloads that was brought up in  https://github.com/pytorch/pytorch/pull/23266.

This PR only compiles one if branch if the condition is an isinstance check. This is consistent with what mypy does; it does not report errors if a branch can be determined statically to be unreachable.

```
from typing import overload

def foo(x):
    # type: (int) -> int
    if isinstance(x, str):
        return x["1"]
    return x + 1

reveal_type(foo) # no error, shows int -> int
```

driazati commented on test/test_jit.py:
```diff
@@ -12803,6 +12803,35 @@ def test({arg_str}):
 
             FileCheck().check_not("prim::PythonOp").run(cu.test.graph)
 
+    def test_isinstance_metacompile(self):
```
This should also check something where the value could be either type, i.e. a class attribute

```python
class M(torch.nn.Module):
	def __init__(self, x):
		self.x = x
	...

torch.jit.script(M(2))
torch.jit.script(M('hi'))
```

eellison replied:
`isinstance` is supposed to throw if it can't be statically resolved (don't have runtime support); that's an existing [limitation](https://github.com/pytorch/pytorch/blob/8fc349f7be84de74fbdadaca518a69af947d6739/torch/csrc/jit/script/compiler.cpp#L2181)

driazati replied:
It is still statically resolved here, each module statically has a different type for that attribute name

Chillee commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -1272,18 +1272,44 @@ struct to_ir {
     // corresponding branches
     Expr cond = stmt.cond();
+    bool isinstance_call = isIsInstanceCall(cond);
+    bool potential_none_check = !isinstance_call && isPotentialNoneCheck(cond);
```
nit: Is `isIsInstanceCall(cond)` ever true when `isPotentialNoneCheck(cond)` is true? Otherwise, seems like this conditional can just be `bool potential_none_check = isPotentialNoneCheck(cond)`.

eellison replied:
This is maybe over-optimization.  `isPotentialNoneCheck(cond)` is never true when `isIsInstanceCall(cond)` is true. So here I am only computing `isPotentialNoneCheck` when `isIsInstanceCall` is false.

If you think i should remove it i will

Chillee commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -1272,18 +1272,44 @@ struct to_ir {
+    if (isinstance_call) {
+      auto is_instance_result = emitSugaredExpr(cond, 1);
+      auto ivalue = toIValue(is_instance_result->asValue(cond.range(), method));
+      TORCH_INTERNAL_ASSERT(ivalue); // no support for runtime checks
```
nit: Shouldn't this be an user facing error? My understanding is that `TORCH_INTERNAL_ASSERT` is reserved for things that should never happen.

eellison replied:
This should never happen. If you look at implementation of `isinstance` it always [returns a constant](https://github.com/pytorch/pytorch/blob/8fc349f7be84de74fbdadaca518a69af947d6739/torch/csrc/jit/script/compiler.cpp#L2196)

Chillee reviewed (approved):
Looks fairly straightforward and logic checks out.

Commit:
Update on "[JIT] metacompile isinstance checks"

This is a series of PRs that will allow us to support adding [padding to conv](https://github.com/pytorch/pytorch/pull/22484) and also reduce the friction of adding method overloads that was brought up in  https://github.com/pytorch/pytorch/pull/23266.

This PR only compiles one if branch if the condition is an isinstance check. This is consistent with what mypy does; it does not report errors if a branch can be determined statically to be unreachable.

```
def foo(x):
    # type: (int) -> int
    if isinstance(x, str):
        return x["1"]
    return x + 1

reveal_type(foo) # no error, shows int -> int
``` 



Differential Revision: [D16697092](https://our.internmc.facebook.com/intern/diff/D16697092)
